tempesta_fw/apm.c

```diff
@@ -1151,14 +1496,318 @@ tfw_cfgop_apm_stats(TfwCfgSpec *cs, TfwCfgEntry *ce)
+
+	return 0;
+}
+
+static int
+tfw_cfgop_apm_hm_request(TfwCfgSpec *cs, TfwCfgEntry *ce)
+{
+	unsigned long size;
+
+	BUG_ON(!tfw_hm_entry);
+	if (tfw_cfg_check_single_val(ce))
+		return -EINVAL;
+
+	size = strlen(ce->vals[0]);
+	tfw_hm_entry->req = (char *)__get_free_pages(GFP_KERNEL,
+						     get_order(size));
+	if (!tfw_hm_entry->req) {
+		TFW_ERR_NL("Can't allocate memory for helth"
+			   " monitoring request\n");
+		return -ENOMEM;
+	}
+	memcpy(tfw_hm_entry->req, ce->vals[0], size);
+	tfw_hm_entry->reqsz = size;
+
+	return 0;
+}
+
+static int
+tfw_cfgop_apm_hm_req_url(TfwCfgSpec *cs, TfwCfgEntry *ce)
+{
+	char *mptr;
+	int size;
+
+	BUG_ON(!tfw_hm_entry);
+	if (tfw_cfg_check_single_val(ce))
+		return -EINVAL;
+
+	size = strlen(ce->vals[0]);
+	mptr = kzalloc(size, GFP_KERNEL);
+	if (!mptr) {
+		TFW_ERR_NL("Can't allocate memory for '%s'\n",
+			   ce->vals[0]);
+		return -ENOMEM;
+	}
+	memcpy(mptr, ce->vals[0], size);
+	tfw_hm_entry->url = mptr;
+	tfw_hm_entry->urlsz = size;
+
+	return 0;
+}
+
+static int
+tfw_cfgop_apm_hm_resp_code(TfwCfgSpec *cs, TfwCfgEntry *ce)
+{
+	int i, code;
+	const char *val;
+
+	if (!ce->val_n) {
+		TFW_ERR_NL("No arguments found.\n");
+		return -EINVAL;
+	}
+	if (ce->attr_n) {
+		TFW_ERR_NL("Unexpected attributes\n");
+		return -EINVAL;
+	}
+	TFW_CFG_ENTRY_FOR_EACH_VAL(ce, i, val) {
+		if (tfw_cfgop_parse_http_status(val, &code)
+		    || tfw_cfg_check_range(code, HTTP_CODE_MIN, HTTP_CODE_MAX))
+		{
+			TFW_ERR_NL("Unable to parse http code value: '%s'\n",
+				   val);
+			return -EINVAL;
+		}
+		__set_bit(HTTP_CODE_BIT_NUM(code), tfw_hm_entry->codes);
+	}
+	return 0;
+}
+
+static int
+tfw_cfgop_apm_hm_resp_crc32(TfwCfgSpec *cs, TfwCfgEntry *ce)
+{
+	u32 crc32;
+	if (tfw_cfg_check_single_val(ce))
+		return -EINVAL;
+
+	if (tfw_cfg_parse_int(ce->vals[0], &crc32)) {
+		TFW_ERR_NL("Unable to parse crc32 value: '%s'\n", ce->vals[0]);
+		return -EINVAL;
+	}
+
+	tfw_hm_entry->crc32 = crc32;
+
+	return 0;
+}
+
+static int
+tfw_cfgop_apm_hm_timeout(TfwCfgSpec *cs, TfwCfgEntry *ce)
+{
+	int timeout;
+
+	if (tfw_cfg_check_single_val(ce))
+		return -EINVAL;
+
+	if (tfw_cfg_parse_int(ce->vals[0], &timeout)) {
+		TFW_ERR_NL("Unable to parse http timeout value: '%s'\n",
+			   ce->vals[0]);
+		return -EINVAL;
+	}
+	if (tfw_cfg_check_range(timeout, 1, USHRT_MAX))
+		return -EINVAL;
+
+	tfw_hm_entry->tmt = timeout;
+
+	return 0;
+}
+
+static void
+tfw_cfgop_cleanup_apm_hm(TfwCfgSpec *cs)
+{
+	TfwApmHM *hm, *tmp;
+
+	if (list_empty(&tfw_hm_list))
+		return;
+
+	list_for_each_entry_safe(hm, tmp, &tfw_hm_list, list) {
+		free_pages((unsigned long)hm->req, get_order(hm->reqsz));
+		kfree(hm->url);
+		list_del(&hm->list);
+		kfree(hm);
+	}
+	INIT_LIST_HEAD(&tfw_hm_list);
+}
+
+static TfwCfgSpec tfw_apm_hm_specs[] = {
+	{
+		.name		= "request",
+		.deflt		= "\"GET / HTTTP/1.0\r\n\r\n\"",
+		.handler	= tfw_cfgop_apm_hm_request,
+		.allow_none	= false,
+		.allow_repeat	= false,
+	},
+	{
+		.name		= "request_url",
+		.deflt		= NULL,
```

Comment:
I believe, that If the `request` option has a default value, then it's logical that this option should have a default value as well.

Reply:
Corrected.